File: ex/hexagonal/src/main/kotlin/com/ktb/shop/domain/port/ProductCatalog.kt
package com.ktb.shop.domain.port

import com.ktb.shop.domain.CategoryId
import com.ktb.shop.domain.Product
import com.ktb.shop.domain.ProductId

// DRIVEN port. With ~1K products it has both a DB adapter and a cached one.
interface ProductCatalog {
    fun byCategory(category: CategoryId): List<Product>
    fun byId(id: ProductId): Product?
    fun all(): List<Product>
}

File: ex/hexagonal-go/internal/domain/port/product_catalog.go
package port

import (
	"context"

	"github.com/ktb/hexagonal-shop/internal/domain"
)

// ProductCatalog is a DRIVEN port. With ~1K products it has both a DB adapter
// and a cached one.
type ProductCatalog interface {
	ByCategory(ctx context.Context, category domain.CategoryID) ([]domain.Product, error)
	ByID(ctx context.Context, id domain.ProductID) (*domain.Product, error)
	All(ctx context.Context) ([]domain.Product, error)
}
